crates/lib/src/bootc_composefs/delete.rs

```diff
@@ -0,0 +1,370 @@
+            continue;
+        }
+
+        buffer.write_all(entry.to_string().as_bytes())?;
+    }
+
+    grub_dir
+        .atomic_write(USER_CFG_STAGED, buffer)
+        .with_context(|| format!("Writing to {USER_CFG_STAGED}"))?;
+
+    rustix::fs::fsync(grub_dir.reopen_as_ownedfd().context("Reopening")?).context("fsync")?;
+
+    rename_exchange_user_cfg(&grub_dir)
+}
+
+fn delete_depl_boot_entries(deployment: &DeploymentEntry, deleting_staged: bool) -> Result<()> {
+    match deployment.deployment.bootloader {
+        Bootloader::Grub => {
+            let boot_dir = Dir::open_ambient_dir("/sysroot/boot", ambient_authority())
+                .context("Opening boot dir")?;
+
+            match deployment.deployment.boot_type {
+                BootType::Bls => delete_type1_entries(deployment, &boot_dir, deleting_staged),
+
+                BootType::Uki => {
+                    let device = get_sysroot_parent_dev()?;
+                    let (esp_part, ..) = get_esp_partition(&device)?;
+                    let esp_mount = TempMount::mount_dev(&esp_part)?;
+
+                    delete_uki(&deployment.deployment.verity, &esp_mount.fd)?;
+
+                    remove_grub_menucfg_entry(
+                        &deployment.deployment.verity,
+                        &boot_dir,
+                        deleting_staged,
+                    )
+                }
+            }
+        }
+
+        Bootloader::Systemd => {
+            let device = get_sysroot_parent_dev()?;
+            let (esp_part, ..) = get_esp_partition(&device)?;
+
+            let esp_mount = TempMount::mount_dev(&esp_part)?;
+
+            // For Systemd UKI as well, we use .conf files
+            delete_type1_entries(deployment, &esp_mount.fd, deleting_staged)
+        }
+    }
+}
+
+pub(crate) async fn delete_composefs_deployment(deployment_id: &str, delete: bool) -> Result<()> {
+    let host = composefs_deployment_status().await?;
+
+    let booted = host.require_composefs_booted()?;
+
+    let all_depls = host.all_composefs_deployments()?;
+
+    let depl_to_del = all_depls
+        .iter()
+        .find(|d| d.deployment.verity == deployment_id);
+
+    let Some(depl_to_del) = depl_to_del else {
+        anyhow::bail!("Deployment {deployment_id} not found");
+    };
+
+    let deleting_staged = host
+        .status
+        .staged
+        .as_ref()
+        .and_then(|s| s.composefs.as_ref())
+        .map_or(false, |cfs| cfs.verity == deployment_id);
+
+    // Get all objects referenced by all images
+    // Delete objects that are only referenced by the deployment to be deleted
+
+    // Unqueue rollback. This makes it easier to delete boot entries later on
+    if matches!(depl_to_del.ty, Some(Slot::Rollback)) && host.status.rollback_queued {
+        composefs_rollback().await?;
+    }
+
+    let sysroot =
+        Dir::open_ambient_dir("/sysroot", ambient_authority()).context("Opening sysroot")?;
+
+    let repo = open_composefs_repo(&sysroot)?;
```

Comment:
I'd still like this stuff to key off Storage but maybe we can just do a mass conversion later